compiler/rmi/rmi-interface/src/org/jetbrains/kotlin/rmi/DaemonParams.kt

```diff
@@ -0,0 +1,313 @@
+//}
+
+// --------------------------------------------------------
+
+open class PropMapper<C, V, P: KMutableProperty1<C, V>>(val dest: C,
+                                                        val prop: P,
+                                                        val names: List<String> = listOf(prop.name),
+                                                        val fromString: (s: String) -> V,
+                                                        val toString: ((v: V) -> String?) = { it.toString() },
+                                                        val skipIf: ((v: V) -> Boolean) = { false },
+                                                        val mergeDelimiter: String? = null)
+{
+    open fun toArgs(prefix: String = COMPILE_DAEMON_CMDLINE_OPTIONS_PREFIX): List<String> =
+            when {
+                skipIf(prop.get(dest)) -> listOf<String>()
+                mergeDelimiter != null -> listOf(prefix + names.first() + mergeDelimiter + toString(prop.get(dest))).filterNotNull()
+                else -> listOf(prefix + names.first(), toString(prop.get(dest))).filterNotNull()
+            }
+    fun apply(s: String) = prop.set(dest, fromString(s))
+}
+
+class StringPropMapper<C, P: KMutableProperty1<C, String>>(dest: C,
+                                                           prop: P,
+                                                           names: List<String> = listOf(),
+                                                           fromString: ((String) -> String) = { it },
+                                                           toString: ((String) -> String?) = { it.toString() },
+                                                           skipIf: ((String) -> Boolean) = { it.isEmpty() },
+                                                           mergeDelimiter: String? = null)
+: PropMapper<C, String, P>(dest = dest, prop = prop, names = if (names.any()) names else listOf(prop.name),
+                            fromString = fromString,  toString = toString, skipIf = skipIf, mergeDelimiter = mergeDelimiter)
+
+class BoolPropMapper<C, P: KMutableProperty1<C, Boolean>>(dest: C, prop: P, names: List<String> = listOf())
+    : PropMapper<C, Boolean, P>(dest = dest, prop = prop, names = if (names.any()) names else listOf(prop.name), 
+                                fromString = { true },  toString = { null }, skipIf = { !prop.get(dest) })
+
+class RestPropMapper<C, P: KMutableProperty1<C, MutableCollection<String>>>(dest: C, prop: P) 
+    : PropMapper<C, MutableCollection<String>, P>(dest = dest, prop = prop, toString = { null }, fromString = { arrayListOf() }) 
+{
+    override fun toArgs(prefix: String): List<String> = prop.get(dest).map { prefix + it }
+    fun add(s: String) { prop.get(dest).add(s) }
+}
+
+// ------------------------------------------
+
+fun Iterable<String>.filterExtractProps(propMappers: List<PropMapper<*,*,*>>, prefix: String, restParser: RestPropMapper<*,*>? = null) : Iterable<String>  {
```

Comment:
During the previous review I slightly rewrite this and something else, I hope this will be useful.
https://github.com/JetBrains/kotlin/commit/ef2ea84a85c3415fc9fa1c933e44c880df979d80#diff-4ff40116c0033e709c32c7abef5b2b0eR84